File: server/pdf/quote_pdf_formatters.py
"""Display formatting for quotation PDF (locale-style, print-safe)."""

from __future__ import annotations

import re
from datetime import date
from decimal import Decimal, ROUND_CEILING, ROUND_HALF_UP


def _dec(x: float | int | Decimal) -> Decimal:
    return Decimal(str(x))


def _decimal_to_plain_str(q: Decimal) -> str:
    """Fixed-point string without scientific notation (avoids 1E+1 from normalize())."""
    t = format(q, "f")
    if "." in t:
        t = t.rstrip("0").rstrip(".")
    return t if t else "0"


def format_currency(amount: float | int | Decimal, currency_code: str) -> str:
    code = (currency_code or "EUR").strip().upper()
    q = _dec(amount).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
    neg = q < 0
    q = abs(q)
    s = f"{q:,.2f}"
    # Thin space after thousands is common in EU; keep comma for readability in PDF
    symbols = {
        "EUR": "€",
        "USD": "$",
        "GBP": "£",
        "ILS": "₪",
    }
    sym = symbols.get(code, f"{code} ")
    if code in symbols and code != "USD":
        return f"-{sym}{s}" if neg else f"{sym}{s}"
    if code == "USD":
        return f"-{sym}{s}" if neg else f"{sym}{s}"
    return f"-{sym}{s}" if neg else f"{sym}{s}"


def format_date_display(value: str) -> str:
    """Accept YYYY-MM-DD or pass through already formatted strings."""
    v = (value or "").strip()
    if re.fullmatch(r"\d{4}-\d{2}-\d{2}", v):
        try:
            d = date.fromisoformat(v)
            return d.strftime("%d %b %Y")
        except ValueError:
            pass
    return v


def format_date_il(value: str) -> str:
    """DD/MM/YYYY for Hebrew quotation PDFs."""
    v = (value or "").strip()
    if re.fullmatch(r"\d{4}-\d{2}-\d{2}", v):
        try:
            d = date.fromisoformat(v)
            return d.strftime("%d/%m/%Y")
        except ValueError:
            pass
    return v


def format_kg(kg: float | int) -> str:
    q = _dec(kg).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
    return f"{q:,.2f} kg"


def format_kg_table_cell(kg: float | int) -> str:
    """Line weight for plate table body: number only, unit in column header."""
    q = _dec(kg).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
    return f"{q:,.2f}"


def format_m2(m2: float | int) -> str:
    q = _dec(m2).quantize(Decimal("0.02"), rounding=ROUND_HALF_UP)
    return f"{q:,.2f} m²"


def format_thickness_mm(mm: float | int) -> str:
    q = _dec(mm).quantize(Decimal("0.1"), rounding=ROUND_HALF_UP)
    return f"{q.normalize()} mm".replace(".0 mm", " mm")


def format_mm_one(mm: float | int) -> str:
    """Single dimension in millimetres (for part breakdown columns)."""
    q = int(round(float(mm)))
    return f"{q:,} mm".replace(",", " ")


def format_mm_one_he(mm: float | int) -> str:
    """Millimetres with Hebrew unit label (for RTL PDF)."""
    q = int(round(float(mm)))
    return f"{q:,} מ״מ".replace(",", " ")


def format_thickness_mm_he(mm: float | int) -> str:
    q = _dec(mm).quantize(Decimal("0.1"), rounding=ROUND_HALF_UP)
    s = _decimal_to_plain_str(q)
    return f"{s} מ״מ"


def format_thickness_mm_table_cell(mm: float | int) -> str:
    """Thickness in mm for plate table body: number only, unit in column header."""
    q = _dec(mm).quantize(Decimal("0.1"), rounding=ROUND_HALF_UP)
    return _decimal_to_plain_str(q)


def format_size_mm(length_mm: float | int, width_mm: float | int) -> str:
    lq = int(round(float(length_mm)))
    wq = int(round(float(width_mm)))
    return f"{lq:,} × {wq:,} mm".replace(",", " ")


def format_qty(qty: int) -> str:
    return f"{qty:,}"


def format_int_comma(n: int) -> str:
    """Integer with thousands comma separators (e.g. 22,499)."""
    return f"{n:,}"


def format_currency_amount_only(amount: float | int | Decimal, currency_code: str) -> str:
    """Numeric part only (no symbol), 2 decimals, thousands commas — for PDF metrics row."""
    q = _dec(amount).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
    neg = q < 0
    q = abs(q)
    s = f"{q:,.2f}"
    return f"-{s}" if neg else s


def format_currency_amount_ceil_int(amount: float | int | Decimal) -> str:
    """Integer amount (round toward +∞), thousands commas — for PDF price metric, no decimals."""
    q = _dec(amount)
    n = int(q.to_integral_value(rounding=ROUND_CEILING))
    return format_int_comma(n)


def currency_symbol(currency_code: str) -> str:
    code = (currency_code or "EUR").strip().upper()
    return {
        "EUR": "€",
        "USD": "$",
        "GBP": "£",
        "ILS": "₪",
    }.get(code, f"{code} ")


def format_metric_kg_one_decimal(kg: float | int) -> str:
    """like finalize strip: one decimal, comma thousands."""
    q = _dec(kg).quantize(Decimal("0.1"), rounding=ROUND_HALF_UP)
    return f"{q:,.1f}"


def format_metric_m2_two_decimals(m2: float | int) -> str:
    q = _dec(m2).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
    return f"{q:,.2f}"

File: lib/server/quotePdf/quotePdfFormatters.ts
/**
 * Mirrors server/pdf/quote_pdf_formatters.py (Hebrew quotation PDF, print-safe).
 */
function dec(x: number): number {
  if (!Number.isFinite(x)) return 0;
  return x;
}

function decimalToPlainStr(n: number): string {
  let t = n.toFixed(10).replace(/\.?0+$/, "");
  if (t === "" || t === "-0") return "0";
  return t;
}

export function formatCurrency(amount: number, currencyCode: string): string {
  const code = (currencyCode || "EUR").trim().toUpperCase();
  const q = Math.round(Math.abs(dec(amount)) * 100) / 100;
  const neg = dec(amount) < 0;
  const s = q.toFixed(2).replace(/\B(?=(\d{3})+(?!\d))/g, ",");
  const symbols: Record<string, string> = {
    EUR: "€",
    USD: "$",
    GBP: "£",
    ILS: "₪",
  };
  const sym = symbols[code] ?? `${code} `;
  if (code in symbols && code !== "USD") {
    return neg ? `-${sym}${s}` : `${sym}${s}`;
  }
  if (code === "USD") {
    return neg ? `-${sym}${s}` : `${sym}${s}`;
  }
  return neg ? `-${sym}${s}` : `${sym}${s}`;
}

export function formatDateIl(value: string): string {
  const v = (value || "").trim();
  if (/^\d{4}-\d{2}-\d{2}$/.test(v)) {
    const [y, m, d] = v.split("-").map((x) => parseInt(x, 10));
    if (m >= 1 && m <= 12 && d >= 1 && d <= 31) {
      return `${String(d).padStart(2, "0")}/${String(m).padStart(2, "0")}/${y}`;
    }
  }
  return v;
}

export function formatKg(kg: number): string {
  const q = Math.round(dec(kg) * 100) / 100;
  return `${q.toFixed(2).replace(/\B(?=(\d{3})+(?!\d))/g, ",")} kg`;
}

export function formatKgTableCell(kg: number): string {
  const q = Math.round(dec(kg) * 100) / 100;
  return q.toFixed(2).replace(/\B(?=(\d{3})+(?!\d))/g, ",");
}

export function formatM2(m2: number): string {
  const q = Math.round(dec(m2) * 100) / 100;
  return `${q.toFixed(2).replace(/\B(?=(\d{3})+(?!\d))/g, ",")} m²`;
}

export function formatMmOneHe(mm: number): string {
  const q = Math.round(dec(mm));
  return `${q.toString().replace(/\B(?=(\d{3})+(?!\d))/g, " ")} מ״מ`;
}

export function formatThicknessMmHe(mm: number): string {
  const q = Math.round(dec(mm) * 10) / 10;
  const s = decimalToPlainStr(q);
  return `${s} מ״מ`;
}

export function formatThicknessMmTableCell(mm: number): string {
  const q = Math.round(dec(mm) * 10) / 10;
  return decimalToPlainStr(q);
}

export function formatQty(qty: number): string {
  return String(Math.floor(qty)).replace(/\B(?=(\d{3})+(?!\d))/g, ",");
}

export function formatIntComma(n: number): string {
  return String(Math.trunc(n)).replace(/\B(?=(\d{3})+(?!\d))/g, ",");
}

export function formatCurrencyAmountOnly(amount: number, currencyCode: string): string {
  void currencyCode;
  const q = Math.round(Math.abs(dec(amount)) * 100) / 100;
  const neg = dec(amount) < 0;
  const s = q.toFixed(2).replace(/\B(?=(\d{3})+(?!\d))/g, ",");
  return neg ? `-${s}` : s;
}

export function formatCurrencyAmountCeilInt(amount: number): string {
  const n = Math.ceil(Math.abs(dec(amount))) * (dec(amount) < 0 ? -1 : 1);
  return formatIntComma(n);
}

export function formatMetricKgOneDecimal(kg: number): string {
  const q = Math.round(dec(kg) * 10) / 10;
  return q.toFixed(1).replace(/\B(?=(\d{3})+(?!\d))/g, ",");
}

export function formatMetricM2TwoDecimals(m2: number): string {
  const q = Math.round(dec(m2) * 100) / 100;
  return q.toFixed(2).replace(/\B(?=(\d{3})+(?!\d))/g, ",");
}
